### ISSUE TITLE
Remove query param entirely?

### ISSUE THREAD

**Issue by jdahdah** [justinmahar/react-use-query-param-string#1]
Hi Justin, this is a really handy hook, thank you for building it! I'm using it in conjunction with a [client-side search](https://github.com/angeloashmore/gatsby-plugin-local-search) in an overlay that remembers the query when you hit the back button. I just ran into a little issue that I was hoping you might have some input on.

Looking at https://github.com/justinmahar/react-use-query-param-string/blob/master/src/hooks/useQueryParamString.ts, it doesn't seem like there is a way to fully remove the query param once it is set via `useQueryParamString()`? I'm trying to do this for a "clear search" button. I tried `setUrlQuery(null)` but that kicked off an endless loop. The only option is to set `setUrlQuery("")`, which means that the URL still contains the `?search=` parameter. It would be great if that entire bit could be removed (including the `?` if there are no other params present). I tried something with `window.history.replaceState()` but that gets tricky when there are other params that need to be preserved. Any ideas? Probably something that would have to be added to the hook?

**Comment by justinmahar** [comment]
Hey jdahdah! You're welcome and thanks for opening an issue about this. I'll look into adding a clean way to clear the value entirely, this is a great suggestion.

**Comment by justinmahar** [comment]
tl;dr: Install the latest version and it should address all of the issues you mentioned!

Let me know how that works for you 💪

---

Alright, I tackled this a bit and I've added a few features.
- Query param is cleared from URL by default when the value is equal to the default. You can keep the value in the URL (previous behavior) by giving the new `clearOnDefault` arg to the hook with a value of `false`.
- The hook returns a 4th array arg, `clear`, which is a function that clears the value from the URL entirely.
- The `?` is now removed from the URL when no query params are present!

### The new signature of the hook is as follows:

#### Signature

`useQueryParamString(key: string, defaultValue: string, clearOnInitial?: boolean) => [value, setValue, initialized, clear]`

##### Props

| Prop           | Type     | Description                                                                                                                 |
| -------------- | -------- | --------------------------------------------------------------------------------------------------------------------------- |
| key            | `string` | Required. The key for the query param.                                                                                      |
| defaultValue   | `string` | Required. Default value when no string is present in the query params.                                                      |
| clearOnDefault | `boolean` | Optional. When true, the query param is cleared from the URL when the current value equals the default value. Default true. |

##### Return

The hook returns an [array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array) containing:

| Index | Name        | Type                     | Description                                                                |
| ----- | ----------- | ------------------------ | -------------------------------------------------------------------------- |
| 0     | value       | `string`                 | The query param string value                                               |
| 1     | setValue    | `(value: string) ⇒ void` | A setter for the query param value                                         |
| 2     | initialized | `boolean`                | True after the value has been read from the query params. False otherwise. |
| 3     | clear       | `() => void`             | Clears the query param from the URL and resets to the default value.       |

---

You can install this update with the following:

```
npm i react-use-query-param-string@latest
```

**Comment by jdahdah** [comment]
Wow, that was really fast! Works exactly as I had in mind. Thank you so much. 🙏

**Comment by justinmahar** [comment]
You got it! Thank you so much for the coffee ☕️ You rock.
